language/bytecode_verifier/bytecode_verifier_tests/src/unit_tests/signature_tests.rs

```diff
@@ -1,14 +1,204 @@
+
+    let mut m = test_module();
+    m.struct_handles[0].type_formals = vec![Kind::All];
+    m.struct_handles.push(StructHandle {
+        module: ModuleHandleIndex::new(0),
+        name: IdentifierIndex::new(0),
+        is_nominal_resource: true,
+        type_formals: vec![],
+    });
+    m.locals_signatures[0] = LocalsSignature(vec![Struct(
+        StructHandleIndex::new(0),
+        vec![Struct(StructHandleIndex::new(1), vec![])],
+    )]);
+    let errors = SignatureChecker::new(&m.freeze().unwrap()).verify();
+    assert!(errors.is_empty());
+}
+
+#[test]
+fn unrestricted_as_resource() {
+    use SignatureToken::*;
+
+    let mut m = test_module();
+    m.struct_handles[0].type_formals = vec![Kind::Resource];
+    m.locals_signatures[0] = LocalsSignature(vec![Struct(StructHandleIndex::new(0), vec![Bool])]);
+    let errors = SignatureChecker::new(&m.freeze().unwrap()).verify();
+    assert!(!errors.is_empty());
+}
+
+#[test]
+fn resource_as_unrestricted() {
+    use SignatureToken::*;
+
+    let mut m = test_module();
+    m.struct_handles.push(StructHandle {
+        module: ModuleHandleIndex::new(0),
+        name: IdentifierIndex::new(0),
+        is_nominal_resource: true,
+        type_formals: vec![],
+    });
+    m.struct_handles[0].type_formals = vec![Kind::Unrestricted];
+    m.locals_signatures[0] = LocalsSignature(vec![Struct(
+        StructHandleIndex::new(0),
+        vec![Struct(StructHandleIndex::new(1), vec![])],
+    )]);
+    let errors = SignatureChecker::new(&m.freeze().unwrap()).verify();
+    assert!(!errors.is_empty());
+}
+
+#[test]
+fn all_as_all_ok() {
+    use SignatureToken::*;
+
+    let mut m = test_module();
+    m.struct_handles[0].type_formals = vec![Kind::All];
+    m.function_signatures[0].type_formals = vec![Kind::All];
+    m.locals_signatures[0] = LocalsSignature(vec![Struct(
+        StructHandleIndex::new(0),
+        vec![TypeParameter(0)],
+    )]);
+    let errors = SignatureChecker::new(&m.freeze().unwrap()).verify();
+    assert!(errors.is_empty());
+}
+
+#[test]
+fn all_as_resource() {
+    use SignatureToken::*;
+
+    let mut m = test_module();
+    m.struct_handles[0].type_formals = vec![Kind::Resource];
+    m.function_signatures[0].type_formals = vec![Kind::All];
+    m.locals_signatures[0] = LocalsSignature(vec![Struct(
+        StructHandleIndex::new(0),
+        vec![TypeParameter(0)],
+    )]);
+    let errors = SignatureChecker::new(&m.freeze().unwrap()).verify();
+    assert!(!errors.is_empty());
+}
+
+#[test]
+fn all_as_unrestricted() {
+    use SignatureToken::*;
+
+    let mut m = test_module();
+    m.struct_handles[0].type_formals = vec![Kind::Unrestricted];
+    m.function_signatures[0].type_formals = vec![Kind::All];
+    m.locals_signatures[0] = LocalsSignature(vec![Struct(
+        StructHandleIndex::new(0),
+        vec![TypeParameter(0)],
+    )]);
+    let errors = SignatureChecker::new(&m.freeze().unwrap()).verify();
+    assert!(!errors.is_empty());
+}
```

Comment:
Is there any reason not to implement these as functional tests? This would be more consistent with the other tests for the bytecode verifier, and also I think the tests will be easier to read/maintain/use across different implementations that way.

Reply:
Good catch! Yeah for some reason I forgot I could do these tests in functional tests. Will port them.